python_modules/dagster/dagster/_core/storage/partition_status_cache.py

```diff
@@ -152,62 +122,55 @@ def _blocking_batch_load(
-        if not self.serialized_materialized_partition_subset:
-            return partitions_def.empty_subset()
-
-        return partitions_def.deserialize_subset(self.serialized_materialized_partition_subset)
-
-    def deserialize_failed_partition_subsets(
-        self, partitions_def: PartitionsDefinition
-    ) -> PartitionsSubset:
-        if not self.serialized_failed_partition_subset:
-            return partitions_def.empty_subset()
-
-        return partitions_def.deserialize_subset(self.serialized_failed_partition_subset)
-
-    def deserialize_in_progress_partition_subsets(
-        self, partitions_def: PartitionsDefinition
-    ) -> PartitionsSubset:
-        if not self.serialized_in_progress_partition_subset:
-            return partitions_def.empty_subset()
-
-        return partitions_def.deserialize_subset(self.serialized_in_progress_partition_subset)
-
-    def get_materialized_subset(
-        self,
-        asset_graph_view: AssetGraphView,
-        asset_key: AssetKey,
-        partitions_def: PartitionsDefinition,
-    ) -> EntitySubset[AssetKey]:
-        value = self.deserialize_materialized_partition_subsets(partitions_def)
-        return EntitySubset(
-            asset_graph_view, key=asset_key, value=_ValidatedEntitySubsetValue(value)
-        )
 
-    def get_failed_subset(
-        self,
-        asset_graph_view: AssetGraphView,
-        asset_key: AssetKey,
-        partitions_def: PartitionsDefinition,
-    ) -> EntitySubset[AssetKey]:
-        value = self.deserialize_failed_partition_subsets(partitions_def)
-        return EntitySubset(
-            asset_graph_view, key=asset_key, value=_ValidatedEntitySubsetValue(value)
-        )
+def _get_partitions_subset(
+    partitions_def: PartitionsDefinition,
+    entity_subset: Optional[SerializableEntitySubset[AssetKey]],
+    serialized_partitions_subset: Optional[str],
+) -> PartitionsSubset:
+    if entity_subset:
+        return entity_subset.subset_value
+    elif serialized_partitions_subset:
+        # backcompat
+        return partitions_def.deserialize_subset(serialized_partitions_subset)
+    else:
+        return partitions_def.empty_subset()
 
-    def get_in_progress_subset(
-        self,
-        asset_graph_view: AssetGraphView,
-        asset_key: AssetKey,
-        partitions_def: PartitionsDefinition,
-    ) -> EntitySubset[AssetKey]:
-        value = self.deserialize_in_progress_partition_subsets(partitions_def)
-        return EntitySubset(
-            asset_graph_view, key=asset_key, value=_ValidatedEntitySubsetValue(value)
-        )
+
+def get_materialized_partitions_subset(
+    cache_value: Optional[AssetStatusCacheValue], partitions_def: PartitionsDefinition
+) -> PartitionsSubset:
+    if not cache_value:
+        return partitions_def.empty_subset()
+    return _get_partitions_subset(
+        partitions_def,
+        cache_value.materialized_subset,
+        cache_value.serialized_materialized_partition_subset,
```

Comment:
ah OK, you are doing what I suggested already by double-writing both formats. So we could roll back if needed after the first release and it would continue reading from the old value.

Reply:
note I'm actually only double-reading -- so rolling forward, no data is lost, but in rolling back, the cache would need to be rebuilt from scratch. I can update this to double write as well (and then we can remove the double-write in a later release to make this safer)